content/introduction/how-ipfs-works.md

```diff
@@ -0,0 +1,88 @@
+
+So, to recap, IPFS lets you to give CIDs to content, and link that content together by generating DAGs using IPLD. Now let's move on to the last piece: how you find and move that content around!
+
+In order to find which peer has certain content (_discovery_), you use a [_distributed hash table_](https://en.wikipedia.org/wiki/Distributed_hash_table), or DHT. **IPFS is essentially a database.** Just like key-value pairs, if you have a peer, it will let you find content, and if you have content, it will let you find you the peer where the content is stored.
+
+The <a hrefm src="https://libp2p.io/">libp2p project</a> is the part of the IPFS ecosystem that provides the DHT and handles peers connecting and talking to each other. It can be used as a tool for other distributed systems as well; just as there is "IPFS-flavored IPLD," there's also "IPFS-flavored libp2p."
+
+Once you know where your content is (ie, which peer or peers are storing the various blocks that make up the content you're after), you use the DHT **again** to find the current location of those peers (_routing_). So, in order to get to content, you use libp2p to query the DHT twice.
+
+You've discovered your content, and you've found the current location(s) of that content — now you need to connect to that content (_exchange_). libp2p again to the rescue! _[Maybe something here about bootstrapping, Bitswap, or Graphsync.]_
+
+What makes the IPFS protocol especially useful for peer to peer connections is _multiplexing_. Traditionally, you open a different connection to each protocol you talk to. Using IPFS, you open just one connection, and you multiplex everything on that. For everything your peers have to talk to each other about, you send a little bit of each thing, and the other end knows how to sort those chunks where they belong.
+
+This is useful because establishing a connection in p2p systems is sometimes difficult depending on where they are. Once you have that connectivity, you can do whatever you need on it. And you only need one socket.
+
+
+## And everything is modular
+
+As you may have noticed from this discussion, the IPFS ecosystem is made up of many modular libraries that support specific parts of any distributed system. You can certainly use any part of the stack independently, or combine them in novel ways!
+
+For example, all of the exchange and routing parts of the stack will let you open a connection, and exchange that data with other peers. And _[insert another example here.]_
+
+
+## Summary
+
+The IPFS ecosystem lets you to give CIDs to content, and link that content together by generating DAGs using IPLD. You can discover content using a DHT that's provided by libp2p, and open a connection to that content and move it around via libp2p as well. All of this is held together by the "middle" of stack, which is linked, unique identifiers; that's the essential part that the IPFS is built on.
```

Comment:
```suggestion
The IPFS ecosystem gives CIDs to content, and link that content together by generating IPLD-Merkle-DAGs. You can discover content using a DHT that's provided by libp2p, and open a connection to any provider of that content and move it around using a multiplexed connection. All of this is held together by the "middle" of stack, which is linked, unique identifiers; that's the essential part that the IPFS is built on.
```

Reply:
Done, except I removed `Merkle` because we haven't introduced that concept yet. I just used `IPLD-DAGs` instead.

Reply:
I've reconsidered my thoughts on this. I'm going to take another pass and see if there's a useful place to add in Merkle as a concept.

Reply:
TODO: _See about adding Merkle DAGs to this explainer.

Reply:
Fixed with 3756a1a and other helpful people's edits re: Merkle-DAGs.